### PR TITLE
Add `disallowUnrecognizedKeys` option

### PR DESCRIPTION
**Pull request by jakemac53**
Fixes https://github.com/dart-lang/json_serializable/issues/184

Added on `JsonSerializable` annotation, defaults to `false` (old behavior). If `true` then it throws an `UnrecognizedKeysException` which contains all the information about the allowed keys, unrecognized keys, etc.

As a part of this I also migrated everything to use blocks instead of fat arrows to simplify the codegen (this requires adding in an additional method call).

~~@kevmoo not sure how crazy you want me to go with testing here - there is a single test right now in the yaml_test.dart file which is more of an e2e_test as its checking for the actual error output.~~

Added a nested object in kitchen sink which uses this.

### CODE REVIEW

File: json_annotation/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,15 @@
+## 0.2.7
+
+* Added `JsonSerializable.allowUnrecognizedKeys`.
```
Comment (kevmoo) [inline json_annotation/CHANGELOG.md]: `disallowUnregognizedKeys` – keep the defaults to `false'

File: json_annotation/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,15 @@
...
+  * Defaults to `true` which maintains the previous behavior.
+  * Throws an `UnrecognizedKeysException` if it finds unrecognized keys in any
+    json maps.
```
Comment (kevmoo) [inline json_annotation/CHANGELOG.md]: JSON map used to create the annotated object.

File: json_annotation/lib/src/allowed_keys_helpers.dart [hunk 3]
```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// Helper function used in generated code when
+/// `JsonAnnotation.allowUnrecognizedKeys` is `false`.
+///
+/// Should not be used directly.
+void $checkAllowedKeys(Map map, Iterable<String> allowedKeys) {
+  var invalidKeys = map.keys.where((k) => !allowedKeys.contains(k));
```
Comment (kevmoo) [inline json_annotation/lib/src/allowed_keys_helpers.dart]: Might as well add `.toList` here. If there are no keys, then you create an empty list – cheap.

Then we avoid hitting an iterator twice – which is something we should model...

File: json_annotation/lib/src/json_serializable.dart [hunk 4]
```diff
@@ -4,6 +4,13 @@
 
 /// An annotation used to specify a class to generate code for.
 class JsonSerializable {
+  /// If `true` (the default), then any unrecognized keys passed to the
```
Comment (kevmoo) [inline json_annotation/lib/src/json_serializable.dart]: Flip name here...

File: json_serializable/lib/src/generator_helper.dart [hunk 5]
```diff
@@ -169,34 +170,31 @@ class _GeneratorHelper {
...
+          _buffer.write(_deserializeForField(accessibleFields[field],
+              checkedProperty: true));
+          _buffer.write(');');
         }
+
+        _buffer.writeln('return val; }');
+
         var fieldKeyMap = new Map.fromEntries(fieldsSetByFactory
             .map((k) => new MapEntry(k, _nameAccess(accessibleFields[k])))
-            .where((me) => me.key != me.value));
```
Comment (kevmoo) [inline json_serializable/lib/src/generator_helper.dart]: Undo this?

File: json_serializable/pubspec.yaml [hunk 6]
```diff
@@ -12,7 +12,7 @@ dependencies:
 
   # Use a tight version constraint to ensure that a constraint on
   # `json_annotation`. Properly constrains all features it provides.
-  json_annotation: '>=0.2.6 <0.2.7'
+  json_annotation: '>=0.2.7 <0.2.8'
```
Comment (kevmoo) [inline json_serializable/pubspec.yaml]: Don't you need a dep override to make tests run?

**Comment by jakemac53** [comment]
Updated @kevmoo PTAL (before you give me more merge conflicts lol)

File: json_annotation/lib/src/allowed_keys_helpers.dart [hunk 7]
```diff
@@ -3,7 +3,7 @@
 // BSD-style license that can be found in the LICENSE file.
 
 /// Helper function used in generated code when
-/// `JsonAnnotation.disallowUnregognizedKeys` is `true`.
+/// `JsonAnnotation.disallowUnrecognizedKeys` is `true`.
```
Comment (kevmoo) [inline json_annotation/lib/src/allowed_keys_helpers.dart]: What's `JsonAnnotation`? You mean `JsonSerializable`?

File: json_annotation/CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,15 @@
+## 0.2.7
+
+* Added `JsonSerializable.disallowUnrecognizedKeys`.
+  * Defaults to `false` which maintains the previous behavior.
```
Comment (kevmoo) [inline json_annotation/CHANGELOG.md]: Remove this line

File: json_annotation/CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,15 @@
...
+  * Throws an `UnrecognizedKeysException` if it finds unrecognized keys in the
```
Comment (kevmoo) [inline json_annotation/CHANGELOG.md]: this and next bullet: Put runtime notes in json_serialiable changelog

File: json_annotation/lib/src/json_serializable.dart [hunk 10]
```diff
@@ -4,6 +4,13 @@
 
 /// An annotation used to specify a class to generate code for.
 class JsonSerializable {
+  /// If `false` (the default), then any unrecognized keys passed to the
+  /// generated FromJson factory will be ignored.
+  ///
+  /// If `true`, any unrecognized keys will be treated as an error.
+  /// ```
```
Comment (kevmoo) [inline json_annotation/lib/src/json_serializable.dart]: remove triple-ticks?

File: json_serializable/CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,7 @@
+## 0.5.6
+
+* Added support for `JsonSerializable.disallowUnrecognizedKeys`.
```
Comment (kevmoo) [inline json_serializable/CHANGELOG.md]: Put runtime info here from json_annotation

Comment (kevmoo) [inline json_serializable/CHANGELOG.md]: Also note that `fromJson` functions now use block syntax

File: json_serializable/pubspec.yaml (hunk 6 above)
Comment (kevmoo) [reply json_serializable/pubspec.yaml]: ack

File: json_serializable/test/kitchen_sink/kitchen_sink.wrapped.g.dart [hunk 12]
```diff
@@ -67,6 +72,7 @@ abstract class _$KitchenSinkSerializerMixin {
   bool get writeNotNull;
   String get string;
   SimpleObject get simpleObject;
+  StrictKeysObject get strictKeysObject;
```
Comment (kevmoo) [inline json_serializable/test/kitchen_sink/kitchen_sink.wrapped.g.dart]: nit: it'd be better to put this in the json_serial_example thingy

...this thing is SOOO big. Just update an existing class w/ the annotation to check behavior...

File: json_annotation/CHANGELOG.md (hunk 8 above)
Comment (jakemac53) [reply json_annotation/CHANGELOG.md]: done

File: json_annotation/CHANGELOG.md (hunk 9 above)
Comment (jakemac53) [reply json_annotation/CHANGELOG.md]: done

File: json_annotation/lib/src/json_serializable.dart (hunk 10 above)
Comment (jakemac53) [reply json_annotation/lib/src/json_serializable.dart]: done

File: json_serializable/CHANGELOG.md (hunk 11 above)
Comment (jakemac53) [reply json_serializable/CHANGELOG.md]: done

Comment (jakemac53) [reply json_serializable/CHANGELOG.md]: done

File: json_serializable/test/kitchen_sink/kitchen_sink.wrapped.g.dart (hunk 12 above)
Comment (jakemac53) [reply json_serializable/test/kitchen_sink/kitchen_sink.wrapped.g.dart]: The point is that it literally tests everything though right? Fwiw adding this here did surface an actual error... and it gives coverage for all the combinations of other options which I think is good.

What is the `json_serial_example` thingy?

Comment (jakemac53) [reply json_serializable/test/kitchen_sink/kitchen_sink.wrapped.g.dart]: going to submit as is for now it should be trivial to make this change as a followup if desired

**Pull request merged**